tests/csapi/device_lists_test.go

```diff
@@ -6,61 +6,315 @@ import (
 
 	"github.com/matrix-org/complement/internal/b"
 	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/docker"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/matrix-org/complement/runtime"
+	"maunium.net/go/mautrix/crypto/olm"
 
 	"github.com/tidwall/gjson"
 )
 
-// TestLocalUsersReceiveDeviceListUpdates tests that users on the same
-// homeserver receive device list updates from other local users, as
-// long as they share a room.
-func TestLocalUsersReceiveDeviceListUpdates(t *testing.T) {
-	// Create a homeserver with two users that share a room
-	deployment := Deploy(t, b.BlueprintOneToOneRoom)
-	defer deployment.Destroy(t)
+// TestDeviceListUpdates tests various flows and checks that:
+//  1. `/sync`'s `device_lists.changed/left` contain the correct user IDs.
+//  2. `/keys/query` returns the correct information after device list updates.
+func TestDeviceListUpdates(t *testing.T) {
+	localpartIndex := 0
+	// generateLocalpart generates a unique localpart based on the given name.
+	generateLocalpart := func(localpart string) string {
+		localpartIndex++
+		return fmt.Sprintf("%s%d", localpart, localpartIndex)
+	}
 
-	// Get a reference to the already logged-in CS API clients for each user
-	alice := deployment.Client(t, "hs1", "@alice:hs1")
-	bob := deployment.Client(t, "hs1", "@bob:hs1")
-
-	// Deduce alice's device ID
-	resp := alice.MustDoFunc(
-		t,
-		"GET",
-		[]string{"_matrix", "client", "v3", "account", "whoami"},
-	)
-	responseBodyBytes := client.ParseJSON(t, resp)
-	aliceDeviceID := gjson.GetBytes(responseBodyBytes, "device_id").Str
-
-	// Bob performs an initial sync
-	_, bobNextBatch := bob.MustSync(t, client.SyncReq{})
-
-	// Alice then updates their device list by renaming their current device
-	alice.MustDoFunc(
-		t,
-		"PUT",
-		[]string{"_matrix", "client", "v3", "devices", aliceDeviceID},
-		client.WithJSONBody(
-			t,
-			map[string]interface{}{
-				"display_name": "A New Device Name",
-			},
-		),
-	)
-
-	// Check that Bob received a device list update from Alice
-	bob.MustSyncUntil(
-		t,
-		client.SyncReq{
-			Since: bobNextBatch,
-		}, func(clientUserID string, topLevelSyncJSON gjson.Result) error {
-			// Ensure that Bob sees that Alice has updated their device list
-			usersWithChangedDeviceListsArray := topLevelSyncJSON.Get("device_lists.changed").Array()
+	// uploadNewKeys uploads a new set of keys for a given client.
+	// Returns a check function that can be passed to mustQueryKeys.
+	uploadNewKeys := func(t *testing.T, user *client.CSAPI) []match.JSON {
+		account := olm.NewAccount()
+		ed25519Key, curve25519Key := account.IdentityKeys()
+
+		ed25519KeyID := fmt.Sprintf("ed25519:%s", user.DeviceID)
+		curve25519KeyID := fmt.Sprintf("curve25519:%s", user.DeviceID)
```

Comment:
I can't find the spaces in question. Which lines look inconsistent?

Reply:
I see this: 
![image](https://user-images.githubusercontent.com/8614563/189734580-4c379c98-dea5-45ce-a6e1-7d9c0719aa18.png)

but I think it's some kind of github web ui bug since it doesn't consistently appear if I refresh? :shrug:. Sorry for noise.